go/engine/per_user_key_background.go

```diff
@@ -0,0 +1,192 @@
+// Note that this engine is long-lived and potentially has to deal with being
+// logged out and logged in as a different user, etc.
+
+package engine
+
+import (
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/libkb"
+	context "golang.org/x/net/context"
+)
+
+type PerUserKeyBackgroundSettings struct {
+	Start    time.Duration
+	WakeUp   time.Duration
+	Interval time.Duration
+	Limit    time.Duration
+}
+
+var PerUserKeyBackgroundDefaultSettings = PerUserKeyBackgroundSettings{
+	// Wait after starting the app
+	Start: 30 * time.Second,
+	// When waking up on mobile lots of timers will go off at once. We wait an additional
+	// delay so as not to add to that herd and slow down the mobile experience when opening the app.
+	WakeUp: 10 * time.Second,
+	// Wait between checks
+	Interval: 1 * time.Hour,
+	// Time limit on each round
+	Limit: 5 * time.Minute,
+}
+
+// PerUserKeyBackground is an engine.
+type PerUserKeyBackground struct {
+	libkb.Contextified
+	sync.Mutex
+
+	args *PerUserKeyBackgroundArgs
+
+	shutdown bool
+	// Function to cancel the background context.
+	// Can be nil before RunEngine exits
+	shutdownFunc context.CancelFunc
+}
+
+type PerUserKeyBackgroundArgs struct {
+	Settings PerUserKeyBackgroundSettings
+
+	// Channels used for testing. Normally nil.
+	testingMetaCh     chan<- string
+	testingRoundResCh chan<- error
+}
+
+// NewPerUserKeyBackground creates a PerUserKeyBackground engine.
+func NewPerUserKeyBackground(g *libkb.GlobalContext, args *PerUserKeyBackgroundArgs) *PerUserKeyBackground {
+	return &PerUserKeyBackground{
+		Contextified: libkb.NewContextified(g),
+		args:         args,
+		shutdownFunc: nil,
+	}
+}
+
+// Name is the unique engine name.
+func (e *PerUserKeyBackground) Name() string {
+	return "PerUserKeyBackground"
+}
+
+// GetPrereqs returns the engine prereqs.
+func (e *PerUserKeyBackground) Prereqs() Prereqs {
+	return Prereqs{}
+}
+
+// RequiredUIs returns the required UIs.
+func (e *PerUserKeyBackground) RequiredUIs() []libkb.UIKind {
+	return []libkb.UIKind{}
+}
+
+// SubConsumers returns the other UI consumers for this engine.
+func (e *PerUserKeyBackground) SubConsumers() []libkb.UIConsumer {
+	return []libkb.UIConsumer{&PerUserKeyUpgrade{}}
+}
+
+// Run starts the engine.
+// Returns immediately, kicks off a background goroutine.
+func (e *PerUserKeyBackground) Run(ectx *Context) (err error) {
+	ctx := ectx.NetContext
+	defer e.G().CTrace(ctx, "PerUserKeyBackground", func() error { return err })()
+
+	// use a new background context with a saved cancel function
+	ctx, cancel := context.WithCancel(context.Background())
+
+	e.Lock()
+	defer e.Unlock()
+
+	e.shutdownFunc = cancel
+	if e.shutdown {
+		// Shutdown before started
+		cancel()
+		e.meta("early-shutdown")
+		return nil
+	}
+
+	// start the loop and return
+	go func() {
+		e.meta("loop-start")
+		err := e.loop(ctx, ectx)
+		if err != nil {
+			e.G().Log.CDebugf(ctx, "PerUserKeyBackground loop error: %s", err)
+		}
+		cancel()
```

Comment:
This is the first time I've used context's features, so I could be wrong here. The docs and examples seem to say that you should always call cancel. Meaning one should call it when actually canceled, but also be make sure it gets called at least once.

[Docs from x/net which we use](https://godoc.org/golang.org/x/net/context#WithCancel)
[Example from mainline](https://golang.org/pkg/context/#WithCancel)